### PR TITLE
Feature flags: Refactor & Simplify variant matching

### PR DESCRIPTION
**Pull request by macobo**
## Changes

This is in preparation for https://github.com/PostHog/posthog/issues/7010. Alongside variant matching and overrides a lot of cruft has crept into feature flags, which makes both using and implementing group analytics support harder.

This PR tries to refactor a few things for a cleaner future.

## How did you test this code?

See unit tests.

### CODE REVIEW

**Review by EDsCODE — APPROVED** [review]
don't have much overarching context here but the refatoring is sound

**Review by rcmarron — APPROVED** [review]
This looks good to me from a functional perspective.

I don't have full context on the refactor, but one non-blocking thought: I wonder if it makes sense to move away from the word "group" entirely in feature flags. `match_group` feels like it may still get confusing with group analytics. Maybe we use the word `conditions` in feature flags? So you could have multiple match conditions for a feature flag?

File: posthog/models/feature_flag.py [hunk 1]
```diff
@@ -199,14 +207,9 @@ def get_active_feature_flags(team: Team, distinct_id: str) -> Dict[str, Union[bo
...
-                continue
-            if len(feature_flag.variants) > 0:
-                variant = feature_flag.get_variant_for_distinct_id(distinct_id)
-                if variant is not None:
-                    flags_enabled[feature_flag.key] = variant
-            else:
-                flags_enabled[feature_flag.key] = True
+            match = feature_flag.matches(distinct_id)
+            if match:
+                flags_enabled[feature_flag.key] = match.variant or True
```
Comment (samwinslow) [inline posthog/models/feature_flag.py:212]: 👍  looks great, thank you for clearing up the double negation

**Review by samwinslow — APPROVED** [review]

**Comment by macobo** [comment]
`conditions` is a good word, thanks!

**Pull request merged**